proof/invariant-abstract/AARCH64/ArchVCPU_AI.thy

```diff
@@ -320,33 +362,95 @@ lemma do_reply_transfer_arch_tcb_at[wp]:
   unfolding do_reply_transfer_def
   by (wpsimp wp: gts_wp split_del: if_split)
 
-crunch send_ipc, send_fault_ipc, receive_ipc, handle_fault, handle_interrupt, handle_vm_fault,
+crunch set_extra_badge, send_ipc
+  for cur_domain[wp]: "\<lambda>s. P (cur_domain s)"
+  and etcbs_of[wp]: "\<lambda>s. P (etcbs_of s)"
+  (wp: crunch_wps transfer_caps_loop_pres simp: crunch_simps)
+
+crunch send_fault_ipc, receive_ipc, handle_fault, handle_interrupt, handle_vm_fault,
          handle_hypervisor_fault, send_signal, do_reply_transfer, cancel_all_ipc,
          cancel_all_signals, unbind_maybe_notification, suspend, deleting_irq_handler,
          unbind_notification
-  for valid_cur_vcpu[wp]: valid_cur_vcpu
-  (wp: valid_cur_vcpu_lift_weak crunch_wps)
+  for cur_domain[wp]: "\<lambda>s. P (cur_domain s)"
+  and etcbs_of[wp]: "\<lambda>s. P (etcbs_of s)"
+  and idle_thread[wp]: "\<lambda>s. P (idle_thread s)"
+  and valid_cur_vcpu[wp]: valid_cur_vcpu
+  (wp: valid_cur_vcpu_lift_weak crunch_wps in_cur_domain_lift_weak thread_set_etcbs hoare_vcg_all_liftE_R dxo_wp_weak
+   simp: crunch_simps)
 
 crunch init_arch_objects, reset_untyped_cap
   for arch_state[wp]: "\<lambda>s. P (arch_state s)"
   (wp: crunch_wps preemption_point_inv unless_wp mapME_x_wp'
    simp: crunch_simps)
 
 crunch invoke_untyped
-  for cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
-  and active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
   (wp: crunch_wps mapME_x_inv_wp preemption_point_inv
    simp: crunch_simps mapM_x_def[symmetric] active_cur_vcpu_of_def)
 
+\<comment> \<open>FIXME: Replace earlier lemmas that don't have the Qs. These do have extra preconditions however,
+          so updating proofs for this might not be completely trivial.\<close>
+lemma delete_objects_etcb_at':
+  "\<lbrace>\<lambda>s. Q (etcb_at P t s) \<and> (t \<notin> {ptr..ptr + 2 ^ bits - 1})\<rbrace>
+   delete_objects ptr bits
```

Comment:
If the more general versions have more preconditions, there would be value in keeping the old ones around. I.e. I'd turn the FIXME into a comment that the precondition is the reason we are keeping the old ones.